src/preprocess/federated_pca.py

```diff
@@ -188,26 +200,50 @@ def read_pca_results(self, node, fold):
         evalues = pd.read_csv(evalues_file, sep='\t', header=None)
         evectors = pd.read_csv(evectors_file, sep='\t', header=0)
 
-        return evectors[evectors.columns[5:]].to_numpy(), evalues[0].to_numpy()
+        return evectors[evectors.columns[5:]].to_numpy(), evalues[0].to_numpy(), evectors['ID'].to_numpy()
 
     def create_plink_eigenvec_allele_file(self, fold, evectors):
         # Take the first 5 columns from one of the nodes to mimic combined plink .eigenvec.allele file
         client_allele_file = os.path.join(
             self.result_folder, self.nodes[0], self.train_foldname_template % fold + '.eigenvec.allele'
         )
 
-        allele = pd.read_csv(client_allele_file, sep='\t', header=0)
-        allele = allele[allele.columns[0:5]]
+        client_allele = pd.read_csv(client_allele_file, sep='\t', header=0)
+        server_allele = client_allele[client_allele.columns[0:5]]
 
         for n in range(self.n_components):
```

Comment:
Why we use this PCA preprocessing and also use normalize_stds: True in FL config? Seems that they should do the same thing

Reply:
Not really. There are two different problems:

1. Different STD for train and test / val datasets. This problem comes from the PCA nature, since the principal components of train and test / val datasets are different. We obtain low amplitude while projecting test / val data onto train PCs. We fix this problem by the **std normalization** procedure.
2. The second problem is that projection matrix, obtained via federated PCA, has normalized components (norm = 1). After multiplying it by the initial sample / allele matrix the result projection values are small. Due to that, our MLP is unable to fit successfully. To fix that, I multiply the projection matrix by some constant factor.

Yes, these two procedures can be done is one place one after another. I prefer to have them separately for consistency between federated and centralized PCA data.